### ISSUE TITLE
Can the underlay option be used from the C++ api

### ISSUE THREAD

**Issue by Sicos1977** [qpdf/qpdf#556]
Hi,

Just one question, can the onderlay option be called somehow from the C++ api. I want to do it from code instead of calling qpdf.exe (on Windows)

https://github.com/qpdf/qpdf/blob/master/include/qpdf/qpdf-c.h

**Comment by jberkenbilt** [comment]
The header you've linked to is the C API, which is a small subset of the C++ API. The C API does not support overlay/underlay. You can do everything from the C++ API that you can do from the CLI, but some things are less straightforward because a little bit too much logic appears in the CLI itself. I have plans for the next major version to include some kind of C++ API for doing almost everything you can do from the command line. I haven't worked out the details yet. My goal is to make the CLI a much thinner wrapper around library functionality by adding some higher level interfaces to the library for doing those kinds of operations.

The main interfaces you would want to use would be [getFormXObjectForPage](https://github.com/qpdf/qpdf/blob/release-qpdf-10.3.2/include/qpdf/QPDFPageObjectHelper.hh#L280) and [placeFormXObject](https://github.com/qpdf/qpdf/blob/release-qpdf-10.3.2/include/qpdf/QPDFPageObjectHelper.hh#L305) from QPDFPageObjectHelper.hh. The [pdf-overlay-page.cc](https://github.com/qpdf/qpdf/blob/release-qpdf-10.3.2/examples/pdf-overlay-page.cc) example illustrates their use. You can also find the code in qpdf.cc that implements overlay and underlay [here](https://github.com/qpdf/qpdf/blob/release-qpdf-10.3.2/qpdf/qpdf.cc#L5159) but there's a whole lot of other stuff going on that will make that code hard to follow, so I recommend looking at the example.

All links in this paragraph are to 10.3.2, which is the latest release at this time. It is likely that there will be a better answer in the future. If a future searcher finds this issue and qpdf 11 or later is out, check release notes to see if there are better ways to do it.

**Comment by Sicos1977** [comment]
Where can I find the C++ api?

**Comment by Sicos1977** [comment]
Is it the QPDF.hh file?

**Comment by jberkenbilt** [comment]
Basically all the header files in qpdf/include that end with .hh are part of the C++ API. QPDF.hh is a good place to start. QPDFObjectHandle.hh is also important. The examples directory contains example code for almost all library features.

**Comment by Sicos1977** [comment]
What you also can do and would make it much easier to use qpdf from other programming languages like in my case from C# is to move all the command line parameters into qpdfxxx.dll (xxx = version number). Only exposing a very simple interface to the outside world that accepts the parameters and keep every difficults objects inside. This way it is much easier to generate bindings from another language. Ghostscript does it also this way and is easy to use from C#. It would be rather easy to implement and would soon be done.

**Comment by Sicos1977** [comment]
Like this https://www.ghostscript.com/doc/current/API.htm

Ghostscript accepts all it's parameters through this method `gsapi_set_param()`

**Comment by jberkenbilt** [comment]
@Sicos1977 That's a really good idea and has the advantage that I can also expose something like that to the C interface. I will definitely incorporate that into my thinking.

**Comment by Sicos1977** [comment]
Also an advantage is that you can add new things to QPDF without changing the interface because you just have to parse the argument options and it avoids that I have to do things like this in C# ... I'm not a fan in calling an exe and parse the output that is written to the console :-)

```c#
public void MergePdfs(string inputFile1, string inputFile2, string outputFile, MemoryStream logStream = null)
        {
            if (logStream != null)
                LogStream = logStream;

            var process = new Process();
            var processStartInfo = new ProcessStartInfo
            {
                FileName = _path,
                Arguments = $"--underlay \"{inputFile1}\" -- \"{inputFile2}\" \"{outputFile}\"" ,
                UseShellExecute = false,
                CreateNoWindow = true,
                ErrorDialog = false,
                WindowStyle = ProcessWindowStyle.Hidden,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                LoadUserProfile = false
            };

            process.StartInfo = processStartInfo;
            process.EnableRaisingEvents = true;

            process.OutputDataReceived += delegate(object _, DataReceivedEventArgs args)
            {
                if (!string.IsNullOrWhiteSpace(args.Data))
                    WriteToLog(args.Data);
            };

            process.ErrorDataReceived += delegate(object _, DataReceivedEventArgs args)
            {
                if (!string.IsNullOrWhiteSpace(args.Data))
                    throw new Exception(args.Data);
            };

            for (var retryCounter = 1; retryCounter < _maxRetries; retryCounter++)
            {
                try
                {
                    WriteToLog($"Processing PDF '{inputFile1}' and '{inputFile2}' with underlay option to merge the pages and writing it to '{outputFile}'");
                    process.Start();
                    process.BeginErrorReadLine();
                    process.BeginOutputReadLine();
                    if(process.WaitForExit(60000))
                        WriteToLog("Merged");
                    else
                        throw new Exception($"Calling QPdf on path '{processStartInfo.FileName}' with arguments '{processStartInfo.Arguments}' timed out after 60 seconds");
                    
                    if (!File.Exists(outputFile))
                    {
                        var message = $"The outputfile '{outputFile}' does not exist";
                        WriteToLog(message);
                        throw new ConversionCriticalException(message);
                    }

                    break;
                }
                catch (Exception exception)
                {
                    WriteToLog($"Could not start the QPdf process due to the following reason: {Utilities.ExceptionHelpers.GetInnerException(exception)}");

                    if (retryCounter < _maxRetries)
                    {
                        Thread.Sleep(retryCounter * _retryDelay);
                    }
                    else
                    {
                        var message = $"Could not merge the files '{inputFile1}' and '{inputFile2}'";
                        WriteToLog(message);
                        throw new ConversionCriticalException(message, exception);
                    }
                }
            }
```

**Comment by Sicos1977** [comment]
Any ideas when you are going to implement this? Just curious.

**Comment by jberkenbilt** [comment]
@Sicos1977 See #593 for an ongoing discussion. I'm hoping to get this in in the next few months. My real hope is to get it in in time for the Ubuntu 22.04 feature freeze mid February, but I don't know if I'll have time.

**Comment by Sicos1977** [comment]
I don't know if something like below would work if I want to consume the library from C#. I think it will make it only more diffiecult. The idea behind the way GhostScript is doing it is that the interface is kept very simple so almost any language can write a wrapper for it that does do the interaction with the QPDF code.

```c
auto q = QPDFOperation()
    .withEmptyInput()
    .withOutputFile("/tmp/a.pdf")
    .beginPages()
    .withInput("1.pdf", "password", "1,3,5")
    .withInput("2.pdf", nullptr)
    .endPages()
    .beginEncryption("user", "owner", 256)
    .withPrint(qpdf_r3p_none)
    .endEncryption();
q.run();
```

**Comment by Sicos1977** [comment]
I hate writing code like this because you don't have much control about what a program is doing but in the case of QPDF I had no other way.

```c#
 var process = new Process();
            var processStartInfo = new ProcessStartInfo
            {
                FileName = _path,
                Arguments = $"--underlay \"{inputFile1}\" -- \"{inputFile2}\" \"{outputFile}\"" ,
                UseShellExecute = false,
                CreateNoWindow = true,
                ErrorDialog = false,
                WindowStyle = ProcessWindowStyle.Hidden,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                LoadUserProfile = false
            };
```

**Comment by jberkenbilt** [comment]
Offering a way to work with command-line arguments is in addition to, not instead of, what's in #593. The other solution is more flexible and may allow some things that can't be done from the CLI, but I can still include a way of doing it from the CLI. Doing that with the current code would be very hard, but after I refactor to make the other proposal work, it should be trivial.

**Comment by jberkenbilt** [comment]
@Sicos1977 I have more or less finished implementing this feature on my work branch. I have some more testing and documentation as well as adding a wrapper to the C API, but the feature of being able to a library call with a null-terminated argv array will be in 10.6 which I should get out in the next few weeks.

**Comment by Sicos1977** [comment]
@jberkenbilt : Thanks for taking the time to implement this, It will make it much easier for me to use your library from C#

**Comment by jberkenbilt** [comment]
Changes are in main as of this morning. I'm hoping to get qpdf 10.6 out in the next or two at the most. If you want a sneak preview, here are some things to look at:

* https://github.com/qpdf/qpdf/blob/main/examples/qpdfjob-c.c
* https://github.com/qpdf/qpdf/blob/main/qpdf/qpdfjob-ctest.c
* https://qpdf.readthedocs.io/en/latest/qpdf-job.html

I don't expect to make any more changes to this before 10.6, but the interface isn't "frozen" until the release.

I hope to add more flexibility to this in the future, but at least for now, you will have a way to do from the C API everything that can be done from the CLI.

**Comment by Sicos1977** [comment]
Thanks, now I need to make some time to write a C# wrapper and release it to nuget so that other people also can use it.

**Comment by jberkenbilt** [comment]
Sounds good. I'll be interested to see how it goes. There are ongoing discussions about making the C API more rust-friendly as well.

**Comment by Sicos1977** [comment]
This implementation with the "command line" arguments will it probably make it much easier for rust to. No more translating all kind of object from one language to another programming language

**Comment by jberkenbilt** [comment]
Indeed, though functionality is limited to what you can do from the command-line, and the qpdf library has much more advanced functionality. I am hoping to add the ability to serialize and deseralize PDF to JSON in a future version. That could also be a big help.

**Comment by Sicos1977** [comment]
As a starting point ... now the rest :-)

https://github.com/Sicos1977/QPdfNet

**Comment by Sicos1977** [comment]
I started building my C# wrapper but have a question. What is better to use the `qpdfjob_run_from_argv` or `qpdfjob_run_from_json` and where do I find these functions? I looked in the C api but could not find them overthere

https://github.com/qpdf/qpdf/blob/master/include/qpdf/qpdf-c.h

**Comment by jberkenbilt** [comment]
These functions are in https://github.com/qpdf/qpdf/blob/main/include/qpdf/qpdfjob-c.h . The json version is more flexible and will become more flexible in the future. Neither one is as flexible as the regular API. At this time, the job API is not capable of object-level manipulation. It can only do what the CLI can do. In qpdf 11, I plan on having bidirectional, lossless JSON conversion which would make it possible to do arbitrary manipulations of the PDF from any language that could call the C API, though even that won't have access to the higher level functions from the C++ API.

**Comment by Sicos1977** [comment]
At the moment I'm using the RunFromJSON method like this

```c#
            using var fileStream = new FileStream("d:\\out.txt", FileMode.Create);
            using var streamWriter = new StreamWriter(fileStream);
            streamWriter.AutoFlush = true;
            System.Console.SetOut(streamWriter);
            System.Console.SetError(streamWriter);
            var result = QPdfApi.Native.RunFromJSON("{ \"version\" }");
```

I expected it to print the version number to the console but got nothing back in return... am I doing something wrong here?

**Comment by jberkenbilt** [comment]
The "help" options, --help, --version, --copyright, etc. are not exposed to the JSON API.

**Comment by Sicos1977** [comment]
ow okey that is why I don't get any output :-)

Can you add some kind of warning to the  method when something is used that is not exposed?

**Comment by jberkenbilt** [comment]
I'd be very surprised if the normal error handling method is not saying something. Make sure you are checking errors. If there is an actual problem, please open a separate issue. I am likely to overlook comments on closed issues unless I respond to them right away. Anyway I am not in a place where I can easily check the code or try something with it, but I would have thought that something would be printed to stderr and the function would return 2. It's possible that I am either misremembering how I handled help options or that there is a bug.

**Comment by Sicos1977** [comment]
Is there a place somewhere where I can find all the options that are available using the CLI? I'm busy building a fluent api in C# and it would be nice to know what is possible

For example

```
  "encrypt": {
    "userPassword": "user",
    "ownerPassword": "owner",
    "256bit": {
      "print": "low"
    }
```

How do I know all the options for `print` ... in your example I see `low` ... are there any other options? ... etc. etc...

**Comment by Sicos1977** [comment]
My goal is to make something like this in C#

```c#
            var job = new Job();
            var result = job.InputFile(@"d:\test.pdf")
                .OutputFile(@"d:\output.pdf")
                .Encrypt("test", "test2", EncryptionStrength.Bits128)
                .Run();
```

**Comment by jberkenbilt** [comment]
I've had the idea to make it possible for qpdf to export sufficient information for this, but I didn't get all the way there on the first iteration. I had had something in mind like being able to output some metadata that could drive a GUI which could present a dynamic form to the user and generate json or CLI. That turns out to be fairly complex because there are a number of constraints that are not obvious just from knowing the structure of the json.

A start would be the output of `qpdf --job-json-help`. That just calls `QPDFJob::job_json_schema_v1()` which is, unfortunately, not exposed to the C API. You could get the information by passing the char* array `["anything", "--job-json-help", nullptr]` to `qpdfjob_run_from_argv` and capturing stdout, but I admit this is hardly a natural API. This would give you a pretty big hint about what to put in your API, but it would not be fully automated. You'd still need some special cases. Coming up with a full specification that would allow automatic generation of multi-language APIs is a big undertaking and not one I intend to tackle any time soon.

Ultimately, for creating a more fully functional wrapper around qpdf, you're much better off working with the C++ API than the C API, but if you are interested in specific enhancements that would make this easier to do from the C API, feel free to open new issues. I can't promise anything. I already have on the "roadmap" the idea of capturing output. Bear in mind that QPDFJob was not something designed into qpdf from the start. It was an afterthought to provide a way to expose CLI functionality from the library, and what's there now is just a first iteration.
